src/main.rs

```diff
@@ -229,10 +234,44 @@ impl ChunkHeader {
 
 impl SendState {
     pub fn send_one(&self, start_time: Instant, out: &mut TcpStream) -> Result<SendResult> {
-        let offset = self.offset.fetch_add(MAX_CHUNK_LEN, Ordering::SeqCst);
+        // By deferring the opening of the file descriptor to this point,
+        // we effectively limit the amount of open files to the amount of send threads.
+        // However, this now introduces the possibility of files getting deleted between
+        // getting listed and their turn for transfer.
+        // A vanishing file is expected, it is not a transfer-terminating event.
+        let mut state = self.state.lock();
+        let (offset, in_fd) = match *state {
+            SendStateInner::Pending { ref fname } => {
+                let res = std::fs::File::open(fname);
+                if let Err(ref e) = res {
+                    if e.kind() == std::io::ErrorKind::NotFound {
+                        return Ok(SendResult::FileVanished);
+                    }
+                }
+                let res = res?;
+                let fd = res.as_raw_fd();
+                *state = SendStateInner::InProgress {
+                    file: res,
+                    offset: 0,
+                };
+                (0, fd)
+            }
+            SendStateInner::Done => {
+                return Ok(SendResult::Done);
+            }
+            SendStateInner::InProgress {
+                ref file,
+                mut offset,
+            } => {
+                offset += MAX_CHUNK_LEN;
+                (offset, file.as_raw_fd())
+            }
+        };
+
         let end = self.len.min(offset + MAX_CHUNK_LEN);
 
         if offset >= self.len || offset >= end {
+            *state = SendStateInner::Done;
             return Ok(SendResult::Done);
         }
```

Comment:
```suggestion
        }
        mem::drop(state);
```

The lock guard by default will live until the end of the scope, which means only one thread will be able to send chunks from this file, it holds the lock even across the `sendfile` syscall. If we drop the lock guard here after we are done with the bookkeeping, then multiple threads can send from the same file in parallel.